source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -50,11 +54,131 @@ void Filter::onRead() {
   Network::ConnectionSocket& socket = cb_->socket();
-  std::string proxy_line;
-  if (!readLine(socket.fd(), proxy_line)) {
+
+  if ((!proxyProtocolHeader_.has_value() && !readProxyHeader(socket.fd())) ||
+      (proxyProtocolHeader_.has_value() && !parseExtensions(socket.fd()))) {
+    // We return if a) we do not yet have the header, or b) we have the header but not yet all
+    // the extension data. In both cases we'll be called again when the socket is ready to read
+    // and pick up where we left off.
+    return;
+  }
+
+  if (proxyProtocolHeader_.has_value()) {
+    // Error check the source and destination fields. Most errors are caught by the address
+    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
+    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
+    // address.
+    const auto remote_version = proxyProtocolHeader_.value().remote_address_->ip()->version();
+    const auto local_version = proxyProtocolHeader_.value().local_address_->ip()->version();
+    if (remote_version != proxyProtocolHeader_.value().protocol_version_ ||
+        local_version != proxyProtocolHeader_.value().protocol_version_) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+    // Check that both addresses are valid unicast addresses, as required for TCP
+    if (!proxyProtocolHeader_.value().remote_address_->ip()->isUnicastAddress() ||
+        !proxyProtocolHeader_.value().local_address_->ip()->isUnicastAddress()) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+
+    // Only set the local address if it really changed, and mark it as address being restored.
+    if (*proxyProtocolHeader_.value().local_address_ != *socket.localAddress()) {
+      socket.setLocalAddress(proxyProtocolHeader_.value().local_address_, true);
+    }
+    socket.setRemoteAddress(proxyProtocolHeader_.value().remote_address_);
+  }
+
+  // Release the file event so that we do not interfere with the connection read events.
+  file_event_.reset();
+  cb_->continueFilterChain(true);
+}
+
+size_t Filter::lenV2Address(char* buf) {
+  const uint8_t proto_family = buf[PROXY_PROTO_V2_SIGNATURE_LEN + 1];
+  size_t len;
+
+  switch ((proto_family & 0xf0) >> 4) {
+  case PROXY_PROTO_V2_AF_INET:
+    len = PROXY_PROTO_V2_ADDR_LEN_INET;
+    break;
+  case PROXY_PROTO_V2_AF_INET6:
+    len = PROXY_PROTO_V2_ADDR_LEN_INET6;
+    break;
+  default:
+    throw EnvoyException("Unsupported V2 proxy protocol address family");
+  }
+  return len;
+}
+
+void Filter::parseV2Header(char* buf) {
+  const int ver_cmd = buf[PROXY_PROTO_V2_SIGNATURE_LEN];
+  uint8_t u = buf[PROXY_PROTO_V2_HEADER_LEN - 2];
+  uint8_t l = buf[PROXY_PROTO_V2_HEADER_LEN - 1];
+  size_t hdr_addr_len = (u << 8) + l;
+
+  if ((ver_cmd & 0xf) == PROXY_PROTO_V2_LOCAL) {
+    // This is locally-initiated, e.g. health-check, and should not override remote address
     return;
```

Comment:
There may be an issue here.  Since you only read in the address info before calling this function and read the extensions as a separate `parseExtensions` function, can you get into a state where a `LOCAL` connection has extensions that are never skipped over?  It's handled for IPv4 and IPv6 because you are referencing `WireHeader.extensions_length_`.

```
The 15th and 16th bytes is the address length in bytes in network endian order.
It is used so that the receiver knows how many address bytes to skip even when
it does not implement the presented protocol. Thus the length of the protocol
header in bytes is always exactly 16 + this value. When a sender presents a
LOCAL connection, it should not present any address so it sets this field to
zero. Receivers MUST always consider this field to skip the appropriate number
of bytes and must not assume zero is presented for LOCAL connections.
```

Reply:
i will add a couple of test cases for LOCAL to be certain of not missing a case.